### ISSUE TITLE
GetProcAddress

### ISSUE THREAD

**Issue by illegal-instruction-co** [upx/upx#727]
This issue tracker is **ONLY** used for requesting sensible features.
Please use [stackoverflow](https://stackoverflow.com) for supporting issues.

## What's the problem (or question)?

mmap upx image crashes on GetProcAddress because of it was hooked from target process. Im not sure because i couldnt understand the logic behind upx but if it is getting imports by GetProcAddress in runtime maybe thats why its crashing.

Edit: I think upx pushes GetProcAddressA and call it but they are hooking GetProcAddressForCaller and it takes LPCSTR. But UPX calling with CSTR at runtime. And it is crashing on istrcmp.

## What should have happened?

Get rid of some API calls. 

## Do you have an idea for a solution?

Walking on PEB and get address of necessary functions. 

## How can we reproduce the issue?

1.
2.
3.
4.

## Please tell us details about your environment.

* UPX version used (`upx --version`):
* Host Operating System and version:
* Host CPU architecture:
* Target Operating System and version:
* Target CPU architecture:

**Comment by markus-oberhumer** [comment]
I fear I have no idea what you are trying to say - please provide some concrete info.

**Comment by illegal-instruction-co** [comment]
OK, ill try to make better explain.

I am trying to compress x64 Windows shared library, and injecting it to a process with using some manual mapping tricks. ( i think it doesnt matters to inject it by using LoadLibrary injection )

And target process has replacement ( byte hooking - basically jumping another memory address and do some stuff with parameters and return the original address )  on GetProcAddressForCaller on KERNELBASE.dll. ( Internal call of GetProcAddress ). So the GetProcAddressForCaller method in target process looks like this:


```
FARPROC WINAPI GetProcAddressStub(HMODULE hModule, LPCSTR lpProcName, void* caller)
{
    if (_stricmp(lpProcName, "blabla") == 0) // checking lpProcName
    {
        // doing some stuff here
    }
    return g_origGetProcAddress(hModule, lpProcName, caller); // Returning to original address of GetProcAddressForCaller
}
```

Whenever i inject my shared library without UPX it works as expected. But if i compress my shared library with UPX the target process crashing at:

if (_stricmp(lpProcName, "blabla") == 0) 

So basically, UPX whenever unpacking itself at runtime somehow calling GetProcAddress ( with wrong parameter type than LPCSTR  ) and crashing.

**Comment by markus-oberhumer** [comment]
Well, why do think this is an UPX bug?

**Comment by ml1050** [comment]
UPX always compresses the imports of the DLLs/EXEs. Then the runtime stub uses LoadLibraryA /GetProcAddressA calls to resolve the original imports of the compressed file. If your "manual injecting" procedure can not handle this, then I'm afraid UPX compressed DLLs  will not work you.

AFAIK, the LPCSTR vs CSTR stuff is not relevant in 32/64 bit architectures.

**Comment by illegal-instruction-co** [comment]
> Well, why do think this is an UPX bug?

I dont think its a bug. As i mentioned on issue, i dont have much knowledge about logic of behind UPX but i thought if UPX uses some API calls at runtime, i thought i can advice the get rid of them.

**Comment by illegal-instruction-co** [comment]
> UPX always compresses the imports of the DLLs/EXEs. Then the runtime stub uses LoadLibraryA /GetProcAddressA calls to resolve the original imports of the compressed file. If your "manual injecting" procedure can not handle this, then I'm afraid UPX compressed DLLs will not work you.
> 
> AFAIK, the LPCSTR vs CSTR stuff is not relevant in 32/64 bit architectures.

Is there a way or option to avoid from import compression ?

**Comment by ml1050** [comment]
Import compression can not be disabled.

**Comment by illegal-instruction-co** [comment]
> Import compression can not be disabled.

As logical or is the current code not allows it ? Im asking that because i wonder if there is a possibility to fork the code and make an cl option to do it.

**Comment by ml1050** [comment]
The current implementation always builds a new import table for the compressed file, because the memory sections of the original import data is usually overwritten with compressed data (because of the in-place decompression feature). After decompression, the stub then fixes everything up. This means that  you can not simply disable the import compression.

**Comment by markus-oberhumer** [comment]
At this time I'm **extremely** reluctant to make any changes to the Windows stubs because of Security/Antivirus considerations, so I fear this is `wontfix`.

**Comment by illegal-instruction-co** [comment]
> The current implementation always builds a new import table for the compressed file, because the memory sections of the original import data is usually overwritten with compressed data (because of the in-place decompression feature). After decompression, the stub then fixes everything up. This means that you can not simply disable the import compression.

Well, with my example stub of GetProcAddressForCaller hook, do you have any idea why should be crashed on _istrcmp ?
If you want i can create an example dll with same hook.

**Comment by ml1050** [comment]
Sorry, I have no idea. But if the problem only happens when you manually map the compressed file, then maybe you are doing something differently/incompatibly than what the PE loader of the windows kernel does and UPX expects.

**Comment by markus-oberhumer** [comment]
I don't think this is an UPX issue, so closing.
